File: docx_parser_converter/docx_to_txt/converters/table_converter.py
"""
Module for converting tables to plain text format.
"""
from docx_parser_converter.docx_parsers.models.table_models import Table, TableRow, TableCell
from docx_parser_converter.docx_to_txt.converters.paragraph_converter import ParagraphConverter


class TableConverter:
    """
    Class to convert tables to plain text format.
    """

    @staticmethod
    def convert_table(table: Table, numbering_schema=None, indent: bool = False) -> str:
        """
        Convert a table to plain text format.

        Args:
            table (Table): The table to convert.
            numbering_schema: The numbering schema to use for paragraphs within the table.
            indent (bool): Whether to apply indentation to paragraphs within the table.

        Returns:
            str: Plain text representation of the table.
        """
        if not table or not table.rows:
            return ""

        table_text = ""
        
        # Add a newline before the table
        table_text += "\n"
        
        # Process each row
        for row_index, row in enumerate(table.rows):
            row_text = TableConverter.convert_row(row, numbering_schema, indent)
            table_text += row_text
            
            # Add a separator line between rows
            if row_index < len(table.rows) - 1:
                table_text += "\n"
        
        # Add a newline after the table
        table_text += "\n"
        
        return table_text

    @staticmethod
    def convert_row(row: TableRow, numbering_schema=None, indent: bool = False) -> str:
        """
        Convert a table row to plain text format.

        Args:
            row (TableRow): The table row to convert.
            numbering_schema: The numbering schema to use for paragraphs within the row.
            indent (bool): Whether to apply indentation to paragraphs within the row.

        Returns:
            str: Plain text representation of the row.
        """
        if not row or not row.cells:
            return ""

        row_text = ""
        
        # Process each cell
        for cell_index, cell in enumerate(row.cells):
            cell_text = TableConverter.convert_cell(cell, numbering_schema, indent)
            
            # Add cell text
            row_text += cell_text
            
            # Add a separator between cells
            if cell_index < len(row.cells) - 1:
                row_text += " | "
        
        return row_text

    @staticmethod
    def convert_cell(cell: TableCell, numbering_schema=None, indent: bool = False) -> str:
        """
        Convert a table cell to plain text format.

        Args:
            cell (TableCell): The table cell to convert.
            numbering_schema: The numbering schema to use for paragraphs within the cell.
            indent (bool): Whether to apply indentation to paragraphs within the cell.

        Returns:
            str: Plain text representation of the cell.
        """
        if not cell or not cell.paragraphs:
            return ""

        cell_text = ""
        
        # Process each paragraph in the cell
        for para_index, paragraph in enumerate(cell.paragraphs):
            para_text = ParagraphConverter.convert_paragraph(paragraph, numbering_schema, indent)
            cell_text += para_text
            
            # Add a space between paragraphs in the same cell
            if para_index < len(cell.paragraphs) - 1:
                cell_text += " "
        
        return cell_text

File: docx_parser_converter_ts/src/docx_to_txt/converters/tableConverter.ts
import type { Table, TableRow, TableCell } from '../../docx_parsers/models/tableModels';
import type { NumberingSchema } from '../../docx_parsers/models/numberingModels';
import { ParagraphConverter } from './paragraphConverter';

/**
 * Converts a table to plain text for TXT output, preserving the logic of the Python implementation.
 */
export class TableConverter {
    /**
     * Convert a table to plain text format.
     * @param table The table to convert.
     * @param numberingSchema The numbering schema to use for paragraphs within the table.
     * @param indent Whether to apply indentation to paragraphs within the table.
     * @returns Plain text representation of the table.
     */
    public static convertTable(table: Table, numberingSchema: NumberingSchema, indent: boolean = false): string {
        if (!table || !table.rows || table.rows.length === 0) {
            return '';
        }
        let tableText = '\n'; // Add a newline before the table
        for (let rowIndex = 0; rowIndex < table.rows.length; rowIndex++) {
            const row = table.rows[rowIndex];
            const rowText = TableConverter.convertRow(row, numberingSchema, indent);
            tableText += rowText;
            if (rowIndex < table.rows.length - 1) {
                tableText += '\n'; // Separator line between rows
            }
        }
        tableText += '\n'; // Add a newline after the table
        return tableText;
    }

    /**
     * Convert a table row to plain text format.
     * @param row The table row to convert.
     * @param numberingSchema The numbering schema to use for paragraphs within the row.
     * @param indent Whether to apply indentation to paragraphs within the row.
     * @returns Plain text representation of the row.
     */
    public static convertRow(row: TableRow, numberingSchema: NumberingSchema, indent: boolean = false): string {
        if (!row || !row.cells || row.cells.length === 0) {
            return '';
        }
        let rowText = '';
        for (let cellIndex = 0; cellIndex < row.cells.length; cellIndex++) {
            const cell = row.cells[cellIndex];
            const cellText = TableConverter.convertCell(cell, numberingSchema, indent);
            rowText += cellText;
            if (cellIndex < row.cells.length - 1) {
                rowText += ' | '; // Separator between cells
            }
        }
        return rowText;
    }

    /**
     * Convert a table cell to plain text format.
     * @param cell The table cell to convert.
     * @param numberingSchema The numbering schema to use for paragraphs within the cell.
     * @param indent Whether to apply indentation to paragraphs within the cell.
     * @returns Plain text representation of the cell.
     */
    public static convertCell(cell: TableCell, numberingSchema: NumberingSchema, indent: boolean = false): string {
        if (!cell || !cell.paragraphs || cell.paragraphs.length === 0) {
            return '';
        }
        let cellText = '';
        for (let paraIndex = 0; paraIndex < cell.paragraphs.length; paraIndex++) {
            const paragraph = cell.paragraphs[paraIndex];
            const paraText = ParagraphConverter.convertParagraph(paragraph, numberingSchema, indent);
            cellText += paraText;
            if (paraIndex < cell.paragraphs.length - 1) {
                cellText += ' '; // Space between paragraphs in the same cell
            }
        }
        return cellText;
    }
} 